library/ssl_cli.c

```diff
@@ -790,125 +869,302 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
+    buf    += session_id_len;
+    buflen -= session_id_len;
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, session id len.: %d",
+                               session_id_len ) );
+    MBEDTLS_SSL_DEBUG_BUF( 3,   "client hello, session id",
+                           ssl->session_negotiate->id,
+                           session_id_len );
 
     /*
      * DTLS cookie
      */
+
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
     if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
     {
         if( ssl->handshake->verify_cookie == NULL )
         {
             MBEDTLS_SSL_DEBUG_MSG( 3, ( "no verify cookie to send" ) );
-            *p++ = 0;
+
+            if( buflen < 1 )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small to hold ClientHello" ) );
+                return( MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+            }
+
+            /* Write length byte */
+            *buf = 0;
+
+            buf    += 1;
+            buflen -= 1;
         }
         else
         {
             MBEDTLS_SSL_DEBUG_BUF( 3, "client hello, cookie",
                               ssl->handshake->verify_cookie,
                               ssl->handshake->verify_cookie_len );
 
-            *p++ = ssl->handshake->verify_cookie_len;
-            memcpy( p, ssl->handshake->verify_cookie,
-                       ssl->handshake->verify_cookie_len );
-            p += ssl->handshake->verify_cookie_len;
+            if( buflen < 1 + ssl->handshake->verify_cookie_len )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small to hold ClientHello" ) );
+                return( MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+            }
+
+            /* Write length byte */
+            *buf = ssl->handshake->verify_cookie_len;
+
+            memcpy( buf + 1,
+                    ssl->handshake->verify_cookie,
+                    ssl->handshake->verify_cookie_len );
+
+            buf    += 1 + ssl->handshake->verify_cookie_len;
+            buflen -= 1 + ssl->handshake->verify_cookie_len;
         }
     }
 #endif
 
     /*
      * Ciphersuite list
      */
-    ciphersuites = ssl->conf->ciphersuite_list[ssl->minor_ver];
 
-    /* Skip writing ciphersuite length for now */
-    n = 0;
-    q = p;
-    p += 2;
+    ciphersuites = ssl->conf->ciphersuite_list[ssl->minor_ver];
 
+    /* Compute ciphersuite list length. It consists of
+     * (1) The ciphersuites the client wants to offer,
+     * (2) potentially the SCSV fallback,
+     * (3) potentially the renegotiation indicator.
+     */
+    ciphersuite_list_len = 0;
     for( i = 0; ciphersuites[i] != 0; i++ )
     {
         ciphersuite_info = mbedtls_ssl_ciphersuite_from_id( ciphersuites[i] );
 
+        if( ssl_validate_ciphersuite( ciphersuite_info, ssl,
+                                      ssl->conf->min_minor_ver,
+                                      ssl->conf->max_minor_ver ) != 0 )
+            continue;
+
+        ciphersuite_list_len += 2;
```

Comment:
Why does the length increase by 2 for each validated ciphersuite?

Reply:
`ciphersuite_list_len` is the total length in bytes of the ciphersuite list, and each ciphersuite is encoded in 2 bytes (see the documentation at its declaration).